### PR TITLE
fix lax_numpy array/asarray performance bug

### PR DESCRIPTION
**Pull request by mattjj**
The issue is that `jax.numpy.array(some_array)` doesn't have a fast path, and instead it does the very slow thing of unpacking the leading axis and packing it back together. That performance bug originates from a [typo in `jax.numpy.array`](https://github.com/google/jax/blob/471c8eba990201eb9f626aef860de06ba09e4681/jax/numpy/lax_numpy.py#L1135-L1136) where `__asarray__` was written instead of `__array__`.

### CODE REVIEW

**Comment by mattjj** [comment]
Travis has informed me that this fix is not as simple as I'd naively expected... :)

**Comment by mattjj** [comment]
Actually, this fix was heading in the wrong direction: we don't want to call `__array__` in this function anyway, since that would turn our DeviceArrays back into regular ndarrays.

The fast path we could hope for is if the argument is a numpy.ndarray or a (nested) list/tuple structure of numpy.ndarrays (rather than, say, DeviceArrays), we could keep things as numpy arrays. That would make this code faster by avoiding the creation of a device array:

```python
def collect_experience(steps, use_numpy):
  all_obs = []
  for i in range(steps):
    obs = np.asarray([1., 2, 3, 3])
    if use_numpy:
      obs = onp.asarray(obs) ###<<<<<
    else:
      obs = np.asarray(obs) ###<<<<<
    all_obs.append(obs)
    
  ret = onp.asarray(all_obs)
  return ret
```

I'm going to close this issue and open a more accurate one.
